### PR TITLE
Support rendering of ghost element when creating elements

### PR DESCRIPTION
**Pull request by martin-fleck-at**
- Introduce new element template module
-- Adding new elements to the diagram based on templates
-- Removing templates again

- Support local bounds calculation for client-only added elements
-- Must trigger RequestBounds for hidden view calculation
-- Must not modify main view until ComputedBounds response is handled
-- Mark ResizeHandles as not to be considered for bounds calculation
-- Fix positioning of decorations for hidden view calculation

- Add ghost element extension to tool palette
-- Optional ghost element field in the trigger node creation action
-- Ghost element is added as feedback to the mouse cursor
-- If ghost element is dynamic, templates (palette items) are reloaded

- Bonus: Re-calculate local bounds for resize behavior

Fixes https://github.com/eclipse-glsp/glsp/issues/1159

---

ghost-element
(Mouse cursor is also shown in the application but not on the video unfortunately)

Can be tested with https://github.com/eclipse-glsp/glsp-server-node/pull/65

### CODE REVIEW

**Review by tortmayr — CHANGES_REQUESTED** [review]
In general the change works really well!
I have a couple of inline comments and suggestions but other than that everything looks good to me.

Could you please make sure that all of you newly introduced functions,classes etc. are added to the main index and that add (high level) jsdoc to them?

While marking the local-only `RequestBounds/ComputedBounds` as server actions works it is not really intuitive and rather confusing. I'm wondering whether we should add the a more generic approach here to explicitly restrict a specific action that is normally handle on both client and server to one target.

File: packages/client/src/features/bounds/glsp-hidden-bounds-updater.ts [hunk 1]
```diff
@@ -72,10 +80,59 @@ export class GLSPHiddenBoundsUpdater extends HiddenBoundsUpdater {
...
+
+    protected override getBounds(elm: Node, element: GModelElement & BoundsAware): Bounds {
+        if (!isSVGGraphicsElement(elm)) {
+            this.logger.error(this, 'Not an SVG element:', elm);
+            return Bounds.EMPTY;
+        }
+        if (elm.tagName === 'g') {
+            for (const child of Array.from(elm.children)) {
+                // eslint-disable-next-line no-null/no-null
+                if (child.getAttribute(ATTR_BBOX_ELEMENT) !== null) {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/bounds/glsp-hidden-bounds-updater.ts:99]: No longer relevant after some refactoring.

File: packages/client/src/base/args-feature.ts [hunk 2]
```diff
@@ -35,3 +35,13 @@ export function isArgsAware(element: GModelElement): element is GModelElement &
 export function hasArgs(element?: GModelElement): element is GModelElement & Required<ArgsAware> {
     return element !== undefined && isArgsAware(element) && element.args !== undefined;
 }
+
+export function ensureArgs(element?: GModelElement): element is GModelElement & Required<ArgsAware> {
+    if (element === undefined || !isArgsAware(element)) {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/base/args-feature.ts:40]: Good idea!

File: packages/client/css/ghost-element.css [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+.ghost-element {
+    /* we are a true ghost so we do not want to be used as a target for any mouse event */
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/css/ghost-element.css:18]: Good idea!

File: packages/client/src/features/bounds/glsp-hidden-bounds-updater.ts [hunk 4]
```diff
@@ -72,10 +80,59 @@ export class GLSPHiddenBoundsUpdater extends HiddenBoundsUpdater {
...
+            return bounds;
+        }
+        // END CUSTOMIZATION
+        return elm.getBBox();
+    }
+
+    protected ignoreHiddenBBoxElements(elm: Element): Disposable {
+        const revert = new DisposableCollection();
+        // eslint-disable-next-line no-null/no-null
+        if (isSVGGraphicsElement(elm) && elm.getAttribute(ATTR_HIDDEN_BBOX_ELEMENT) !== null) {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/bounds/glsp-hidden-bounds-updater.ts:128]: No longer relevant after some refactoring.

File: packages/client/src/features/bounds/set-bounds-feedback-command.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    export function is(object: any): object is SetBoundsFeedbackAction {
+        return Action.hasKind(object, KIND);
+    }
+
+    export function create(bounds: ElementAndBounds[]): SetBoundsFeedbackAction {
+        return { kind: KIND, bounds };
+    }
+}
+
+export interface SetBoundsFeedbackAction extends Omit<SetBoundsAction, 'kind'> {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/bounds/set-bounds-feedback-command.ts:43]: Done.

File: packages/client/src/features/decoration/decoration-placer.ts [hunk 6]
```diff
@@ -26,8 +26,8 @@ export class GlspDecorationPlacer extends DecorationPlacer {
         }
         if (isSizeable(element)) {
             return {
-                x: GlspDecorationPlacer.DECORATION_OFFSET.x * element.bounds.width,
-                y: GlspDecorationPlacer.DECORATION_OFFSET.y * element.bounds.height
+                x: -GlspDecorationPlacer.DECORATION_OFFSET.x,
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/decoration/decoration-placer.ts:29]: During the main view rendering, the element bounds are all `-1` so the result was always to position the decoration in the top left corner. However, for the local bounds computation the bounds were already correctly calculated, e.g., 200x100, so the decoration was placed way to the bottom right. I'm not sure why we multiply it with width/height anyway but if there is a particular reason you are aware of, please let me know.

File: packages/client/src/features/element-template/add-template-element.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        id: getTemplateElementId(element.id),
+        cssClasses: [...(element.cssClasses || [])],
+        position: isBoundsAware(element) ? { x: element.bounds.x, y: element.bounds.y } : undefined,
+        size: isBoundsAware(element) ? { width: element.bounds.width, height: element.bounds.height } : undefined,
+        text: element instanceof GLabel ? element.text : undefined,
+        children: element instanceof GParentElement ? element.children.map(child => toElementSchema(child)) : undefined
+    };
+}
+
+export function applyCssClasses(source?: string[], toAdd?: string[], toRemove?: string[]): string[] {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/element-template/add-template-element.ts:108]: Absolutely correct. I kept the function local and added all the other exports to the index, thank you catching this!

File: packages/client/src/features/tool-palette/tool-palette.ts [hunk 8]
```diff
@@ -333,11 +337,15 @@ export class ToolPalette extends AbstractUIExtension implements IActionHandler,
...
 
-    handle(action: EnableDefaultToolsAction): ICommand | Action | void {
+    handle(action: Action): ICommand | Action | void {
         this.changeActiveButton();
-        if (this.focusTracker.hasFocus) {
-            // if focus was deliberately taken do not restore focus to the palette
-            this.focusTracker.diagramElement?.focus();
+        if (UpdateModelAction.is(action) || SetModelAction.is(action)) {
+            this.reloadPaletteBody();
+        } else {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/tool-palette/tool-palette.ts:344]: I went for the type check, thanks!

File: packages/client/src/features/tool-palette/tool-palette.ts [hunk 9]
```diff
@@ -360,11 +368,6 @@ export class ToolPalette extends AbstractUIExtension implements IActionHandler,
         }
     }
 
-    protected handleSetContextActions(action: SetContextActions): void {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/tool-palette/tool-palette.ts:363]: Good idea, I added it!

File: packages/client/src/features/tool-palette/tool-palette.ts [hunk 10]
```diff
@@ -395,16 +398,41 @@ export class ToolPalette extends AbstractUIExtension implements IActionHandler,
...
             editorContext: {
                 selectedElementIds: []
             }
         });
         const response = await this.actionDispatcher.request<SetContextActions>(requestAction);
-        this.paletteItems = response.actions.map(e => e as PaletteItem);
-        if (!this.editorContext.isReadonly) {
-            this.show(this.editorContext.modelRoot);
+        this.paletteItems = response.actions.map(action => action as PaletteItem);
+        this.dynamic = !!this.paletteItems.find(item => this.hasDynamicAction(item));
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/tool-palette/tool-palette.ts:420]: Done.

File: packages/client/src/features/tool-palette/tool-palette.ts [hunk 11]
```diff
@@ -395,16 +398,41 @@ export class ToolPalette extends AbstractUIExtension implements IActionHandler,
...
+    }
+
+    protected hasDynamicAction(item: PaletteItem): boolean {
+        const dynamic = !!item.actions.find(action => TriggerNodeCreationAction.is(action) && action.ghostElement?.dynamic);
+        if (dynamic) {
+            return dynamic;
+        }
+        return !!item.children?.find(child => this.hasDynamicAction(child));
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/tool-palette/tool-palette.ts:428]: Done.

File: packages/client/src/features/tools/node-creation/node-creation-tool.ts [hunk 12]
```diff
@@ -39,14 +45,29 @@ export class NodeCreationTool extends BaseCreationTool<TriggerNodeCreationAction
...
 
     get id(): string {
         return NodeCreationTool.ID;
     }
 
     doEnable(): void {
+        let trackingListener: MouseTrackingElementPositionListener | undefined;
+        const ghostElement = this.triggerAction.ghostElement;
+        if (ghostElement) {
+            trackingListener = new MouseTrackingElementPositionListener(getTemplateElementId(ghostElement.template), this, 'middle');
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/tools/node-creation/node-creation-tool.ts:58]: Done!

File: packages/client/src/utils/layout-utils.ts [hunk 13]
```diff
@@ -18,22 +18,22 @@ import {
...
+    GModelElement,
     ModelLayoutOptions,
     Point,
-    GModelElement,
     Writable
 } from '@eclipse-glsp/sprotty';
 import { IMovementRestrictor } from '../features/change-bounds/movement-restrictor';
 
-export function minWidth(element: GModelElement & BoundsAware): number {
+export function minWidth(element: GModelElement): number {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/utils/layout-utils.ts:28]: I think I needed it for the ghost element at some point but I no longer do. So I reverted those changes.

**Comment by martin-fleck-at** [comment]
@tortmayr I agree that this would be a nice feature to have. However, currently there is no way to simply reload trigger actions and then to uniquely identify them so we can easily replace the action and template that we are currently using. So I pulled this out into an enhancement issue (https://github.com/eclipse-glsp/glsp/issues/1161) as it would require more work that I want to consider out of scope for this ticket. What do you think?

**Review by tortmayr — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**